Source/CarthageKit/DependencyCrawler.swift

```diff
@@ -0,0 +1,226 @@
+	private var dependencyMappings: [Dependency: Dependency]?
+	private let eventPublisher: Signal<DependencyCrawlerEvent, NoError>.Observer
+
+	/// DependencyCrawlerEvent
+	public let events: Signal<DependencyCrawlerEvent, NoError>
+
+	private enum DependencyCrawlerError: Error {
+		case versionRetrievalFailure(message: String)
+		case dependencyRetrievalFailure(message: String)
+	}
+
+	/**
+	Initializes with implementations for retrieving the versions, transitive dependencies and git references.
+
+	Uses the supplied local dependency store to store the encountered dependencies.
+
+	Optional mappings may be specified to anonymize the encountered dependencies (thereby removing sensitive information).
+
+	If ignoreErrors is true, any error during retrieval of the dependencies will not be fatal but will result in an empty array instead.
+	*/
+	public init(
+			versionsForDependency: @escaping (Dependency) -> SignalProducer<PinnedVersion, CarthageError>,
+			dependenciesForDependency: @escaping (Dependency, PinnedVersion) -> SignalProducer<(Dependency, VersionSpecifier), CarthageError>,
+			resolvedGitReference: @escaping (Dependency, String) -> SignalProducer<PinnedVersion, CarthageError>,
+			store: LocalDependencyStore,
+			mappings: [Dependency: Dependency]? = nil,
+			ignoreErrors: Bool = false) {
+		self.versionsForDependency = versionsForDependency
+		self.dependenciesForDependency = dependenciesForDependency
+		self.resolvedGitReference = resolvedGitReference
+		self.store = store
+		self.dependencyMappings = mappings
+		self.ignoreErrors = ignoreErrors
+
+		let (signal, observer) = Signal<DependencyCrawlerEvent, NoError>.pipe()
+		events = signal
+		eventPublisher = observer
+	}
+
+	/**
+	Recursively traverses the supplied dependencies taking into account their compatibleWith version specifiers.
+
+	Stores all dependencies in the LocalDependencyStore.
+
+	Returns a dictionary of all encountered dependencies with as value a set of all their encountered versions.
+	*/
+	public func traverse(dependencies: [Dependency: VersionSpecifier]) -> Result<[Dependency: Set<PinnedVersion>], CarthageError> {
+		let result: Result<[Dependency: Set<PinnedVersion>], CarthageError>
+		do {
+			var handledDependencies = Set<PinnedDependency>()
+			var cachedVersionSets = [Dependency: [PinnedVersion]]()
+			try traverse(dependencies: Array(dependencies),
+						 handledDependencies: &handledDependencies,
+						 cachedVersionSets: &cachedVersionSets)
+			result = .success(handledDependencies.dictionaryRepresentation)
+		} catch let error as CarthageError {
+			result = .failure(error)
+		} catch {
+			result = .failure(CarthageError.internalError(description: error.localizedDescription))
+		}
+		return result
+	}
+
+	private func traverse(dependencies: [(Dependency, VersionSpecifier)],
+						  handledDependencies: inout Set<PinnedDependency>,
+						  cachedVersionSets: inout [Dependency: [PinnedVersion]]) throws {
```

Comment:
We shouldn't use `throws`. Carthage uses typed errors. This should use `Result` with a concrete error type.

Reply:
Is it ok to keep the throws for private methods since it results in a more complicated flow when using Result (multiple exit points that I need to guard against, which require the same handler). These methods are not seen or used from the outside. I did change all public signatures (also in LocalDependencyStore) to use result instead of throws.

Reply:
No, we'd like to use `Result` everywhere for the additional type safety.